synapse/storage/databases/main/deviceinbox.py

```diff
@@ -704,15 +711,15 @@ def _remove_deleted_devices_from_device_inbox_txn(
                     },
                 )
 
-            return num_deleted
+            return num_deleted, stop >= max_stream_id
```

Comment:
`num_deleted` is the wrong number here. We're interested in the number of `stream_id`s we considered (ie, `batch_size`).

Reply:
Oh? As an admin if I read in my logs that an update that deletes stuff from my db "Processed X items" I understand that it deleted X items (not that it deleted items that have a stream ID in a range of size X).

Reply:
the *main* purpose of this value is to feed back into the rate-control algorithm. (The logging is a side-effect).

For the rate-control algorithm to work correctly, we need to be consistent about what sort of thing we are counting. At https://github.com/matrix-org/synapse/pull/11421/files#diff-f0d3815e503803dc732990d65d3d49bcdd08bb60d13c43bff0375cdc06b3538fR676, we're using it to count stream_ids; we therefore need to tell it how many stream_ids we processed, not how many items that resulted in us deleting.

Putting this another way: the RC algorithm will starts us off with a batch size of 100. We consider 100 stream ids, and suppose we end up deleting 1000 inbox entries, all of which takes us 1 second. The RC algorithm is aiming for runtimes of 100ms, so really we should only have done a tenth of that work.
 * if we tell it we deleted 1000 rows in that one second, then next time it will tell us to process (1000/10) items = 100 items again. No improvement.
 * if we tell it we processed 100 stream ids in that one second, then next time it will tell us to process (100/10) items = 10 items - which will be about right, given these hypothetical numbers.

---

If your next question is "what's the point of returning the item count when it should always be the same as the input batch size", well, that's a good question. Sometimes it's hard to break down work at exactly the batch size, so you might end up processing a few extra rows, and returning the number we actually processed is better for the RC algorithm.

Reply:
> At https://github.com/matrix-org/synapse/pull/11421/files#diff-f0d3815e503803dc732990d65d3d49bcdd08bb60d13c43bff0375cdc06b3538fR676, we're using it to count stream_ids; we therefore need to tell it how many stream_ids we processed, not how many items that resulted in us deleting.

Right, in this case that's a very confusing way of doing it - without your explanation I would find it very obvious that the intent is to count the rows that are deleted, which happens to be the same number as `batch_size`.

But it makes sense now, I'll update the code.

Reply:
well, ideas about how to make this clearer are of course welcome...